libpromises/evalfunction.c

```diff
@@ -8451,6 +8451,230 @@ static FnCallResult FnCallNetworkConnections(EvalContext *ctx, ARG_UNUSED const
+                thread_data.path, GetErrorStr());
+            return FnReturnContext(false);
+        }
+        else if (read(fd, buf, sizeof(buf)) < 0)
+        {
+            Log(LOG_LEVEL_DEBUG, "Failed to read from file '%s': %s",
+                thread_data.path, GetErrorStr());
+            close(fd);
+            return FnReturnContext(false);
+        }
+
+        close(fd);
+        return FnReturnContext(true);
+    }
+    // Else try read in separate thread, possibly blocking for N seconds
+    Log(LOG_LEVEL_DEBUG,
+        "Checking if file '%s' is readable in separate thread, "
+        "possibly blocking for %ld seconds.", thread_data.path, timeout);
+
+    struct timespec timespec = {0};
+    xclock_gettime(CLOCK_REALTIME, &timespec);
+    timespec.tv_sec += timeout;
+
+    int ret = pthread_mutex_init(&(thread_data.mutex), NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to initialize mutex: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    // Keep thread from doing its thing until we call
+    // pthread_cond_timedwait(3)
+    ret = pthread_mutex_lock(&(thread_data.mutex));
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to lock mutex: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    ret = pthread_create(&(thread_data.thread), NULL,
+                         &IsReadableThreadRoutine, &thread_data);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to create thread: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    ret = pthread_cond_init(&(thread_data.cond), NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to initialize condition: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    // Wait on thread to finnish or timeout
+    ret = pthread_cond_timedwait(&(thread_data.cond), &(thread_data.mutex),
+                                 &timespec);
+    switch (ret)
+    {
+    case 0: // Thread finnished
+        {
+            ret = pthread_join(thread_data.thread, NULL);
+            if (ret != 0)
+            {
+                Log(LOG_LEVEL_ERR, "Failed to join thread: %s",
+                    GetErrorStrFromCode(ret));
+                return FnFailure();
+            }
+            return thread_data.result;
+        }
+
+    case ETIMEDOUT: // Thread timed out
+        {
+            Log(LOG_LEVEL_DEBUG, "File '%s' is not readable: "
+                "Read operation timed out, exceeded %ld seconds.",
+                thread_data.path, timeout);
+
+            ret = pthread_cancel(thread_data.thread);
+            if (ret != 0)
+            {
+                Log(LOG_LEVEL_ERR, "Failed to cancel thread : %s",
+                    GetErrorStrFromCode(ret));
+                return FnFailure();
+            }
+
+            ret = pthread_join(thread_data.thread, NULL);
```

Comment:
Can't this hang forever if the thread is hanging and cannot be cancelled?

Reply:
>  Does not work with pthread_detach either. 

What if you create the thread as detached? We don't need to `join()` it, right? It either signals the condition in time or we don't care about it.